xlators/storage/posix/src/posix-io-uring.c

```diff
@@ -0,0 +1,634 @@
+/*
+   Copyright (c) 2020 Red Hat, Inc. <http://www.redhat.com>
+   This file is part of GlusterFS.
+
+   This file is licensed to you under your choice of the GNU Lesser
+   General Public License, version 3 or any later version (LGPLv3 or
+   later), or the GNU General Public License, version 2 (GPLv2), in all
+   cases as published by the Free Software Foundation.
+*/
+
+#include "posix.h"
+#include "posix-messages.h"
+#include "posix-io-uring.h"
+#include "posix-handle.h"
+
+#ifdef HAVE_LIBURING
+#include <liburing.h>
+
+struct posix_uring_ctx;
+typedef void(fop_unwind_f)(struct posix_uring_ctx *, int32_t);
+typedef void(fop_prep_f)(struct io_uring_sqe *sqe, struct posix_uring_ctx *);
+static int
+posix_io_uring_submit(xlator_t *this, struct posix_uring_ctx *ctx);
+
+struct posix_uring_ctx {
+    call_frame_t *frame;
+    struct iatt prebuf;
+    dict_t *xdata;
+    fd_t *fd;
+    int _fd;
+    int op;
+
+    union {
+        struct {
+            struct iovec *iov;
+            int count;
+            off_t offset;
+        } write;
+
+        struct {
+            struct iobuf *iobuf;
+            struct iovec iovec;
+            off_t offset;
+        } read;
+
+        struct {
+            int32_t datasync;
+        } fsync;
+    } fop;
+
+    fop_prep_f *prepare;
+    fop_unwind_f *unwind;
+};
+
+static void
+posix_io_uring_ctx_free(struct posix_uring_ctx *ctx)
+{
+    GF_ASSERT(ctx);
```

Comment:
I've just seen that this function can be called with `ctx == NULL`. I would prevent that case, but otherwise we need to remove the `GF_ASSERT()` and check for NULL.

Reply:
Ugh, my bad, this is the reason why I added `if(!ctx) return.` in the earlier version. I will add it again, as it is cleaner than adding `if (ctx) posix_io_uring_ctx_free(ctx)` everywhere.